_posts/2020-07-29-springboot-logging.md

```diff
@@ -0,0 +1,336 @@
+2. Headers: We log important request and response headers, for example, the JWT authorization header, headers giving context information like model of the device that sent the request, geography of the user.
+3. Method entry/exit: Similar to API, we might log the inputs and output of important methods to ensure that all the statements within the method executed successfully.
+4. Useful checkpoints within our source code to indicate the occurrence of any important event, for example, payment successful with payment reference, the message posted to a queue, etc.
+5. Exceptions are often captured in error logs with a complete stack trace and are a valuable component for any system diagnosis. 
+
+### Add Contextual Information
+Logs become even more useful if we add some contextual information to every message we log, for example, the identifier of the user or the name of the business function. Instead of manually appending the contextual information to each log message, we can use Logback’s Mapped Diagnostic Context (MDC). Whatever we put in the MDC can be used in the log pattern. This behavior is comparable to a ThreadLocal variable.
+For example, we put the user idenntifier and function name in the MDC for adding to the log message:
+
+```java
+  @GetMapping("/users/{userID}")
+  public String getUser(@PathVariable("userID") final String userID) {
+    
+    MDC.put("user", userID);
+    MDC.put("function", "userInquiry");
+    
+    logger.info("Controller: Fetching user with id {}", userID);
+    
+    MDC.remove("user");
+    MDC.remove("function");
+    
+    return userService.getUser(userID);
+                                                 
+```
+
+The user identifier and function name in the MDC are then added to the log message in the pattern attribute :
+```
+%d{HH:mm:ss.SSS} [%X{user}] [%X{function}] [%thread] %-5level %logger{36} - %msg%n
+```
+A sample of the log output with the MDC attributes user and function name :
+```shell
+...            : Controller: Fetching user with id john123
+... [john123] [userInquiry] ... - Controller: Fetching user with id john123
+...       : Service: Fetching user with id john123
+... [john123] [userInquiry] ... - Service: Fetching user with id john123
+```
+
+### Different Logging For Each Environment
+
+We often have different logging formats for local and production runtime environments. Spring profiles are an elegant way to implement different logging for each environment. You can refer to a very good use case in this article for (environment-specific logging)[https://reflectoring.io/profile-specific-logging-spring-boot/]. 
+
+### Tracing Requests Across Microservices
+Debugging and tracing in microservice applications is challenging since
+the microservices are deployed and run independently resulting in their logs being distributed across many individual components. We apply two techniques to help simplify the log analysis process of multiple microservices : 
+
+#### Log Aggregation 
```

Comment:
Link this to `logback-spring.xml` and move this section into a subsection of "Customizing the Logger in Spring Boot`